AESinkPipewire: Properly identify HDMI devices

fritsch opened the pull request:
Without this AE would only allow passthrough on SPDIF devices as it checks for the device not being PCM when allowing Passthrough.

fritsch committed:
AESinkPipewire: Properly identify HDMI devices

Without this AE would only allow passthrough on SPDIF devices as it checks
for the device not being PCM when allowing Passthrough.

Co-authored-by: Lukas Rusak <lorusak@gmail.com>

fritsch commented:
@yol FYI

lrusak commented on xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp:
```diff
@@ -348,10 +348,17 @@ void CAESinkPipewire::EnumerateDevicesEx(AEDeviceInfoList& list, bool force)
     }
 
-    if (device.m_channels.Count() == 2 && !device.m_streamTypes.empty())
+    if (!device.m_streamTypes.empty)
```
```suggestion
    if (!device.m_streamTypes.empty())
```

fritsch committed:
AESinkPipewire: Fix Passthrough channel availability

It seems the Pipewire can easily configure TrueHD as IEC Format while
only two channels are available. This is not correct. As later on when AE
starts playing 8 channels are opened, already enumerate properly.

Co-authored-by: Lukas Rusak <lorusak@gmail.com>
Co-authored-by: Markus Härer <markus.haerer@gmx.net>

lrusak commented on xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp:
```diff
@@ -348,10 +348,24 @@ void CAESinkPipewire::EnumerateDevicesEx(AEDeviceInfoList& list, bool force)
+    // If DTS-HD-MA or TrueHD are configured 8 channels are needed
+    bool hasHBRFormat = false;
+    AEDataTypeList::iterator iit, iit2;
+    iit = find(device.m_streamTypes.begin(), device.m_streamTypes.end(), CAEStreamInfo::STREAM_TYPE_TRUEHD);
```
std::find_if would be better here so we don't need to use the temporary iterators.

fritsch replied:
Also with find_if I have to check the return value being not end() right? I don't see how that gets easier? What do I miss?

lrusak replied:
something like this I think

```c++
bool hasHBRFormat = std::find_if(device.m_streamTypes.cbegin(), device.m_streamTypes.cend(), [](const auto& streamType)
{
  if (streamType == CAEStreamInfo::STREAM_TYPE_TRUEHD || streamType == CAEStreamInfo::STREAM_TYPE_DTSHD)
    return true;

  return false;
}) != device.m_streamTypes.cend(); 
```

sky42src commented:
For me there is no change in LibreELEC 12 master + this PR tested with Generic (build RPi4/5 too) in a PipeWire only config. Only PCM 2.0 at the AVR. In playback audio options passthrough on or off does not change anything. No GUI System Audio passthrough options.  
source tag https://github.com/sky42src/LibreELEC.tv/tree/11.80.5-240221-PipeWire  
image donwload https://sky42.libreelec.tv/testing/12.0-PipeWire/  
RPi5 still builds and i will upload when done.

sky42src commented:
@lrusak Any hint for PipeWire config in LE12? As far as i can tell the pipewire support in LE was introduced by you.  I disabled alsa, pulse and enabled pipewire in options.  

@fritsch Do you have any hint for me how to configure correctly a setup with PipeWire only for Kodi with passthrough? I only find some informations with pipewire + alsa or pulseaudio, but no where pure pipewire config.  

Still just PCM 2.0.  
What is the correct setting for
`<setting id="audiooutput.passthroughdevice" default="true">Default</setting>` and how do i find the correct device name? For pipewire there is no GUI passthrough config, so hopefully the right guisettings.xml is enough.

test image update to last version of PR  
source tag https://github.com/sky42src/LibreELEC.tv/tree/11.80.5-240221a-PipeWire  
image donwload https://sky42.libreelec.tv/testing/12.0-PipeWire/

lrusak commented:
@sky42src you need to look at the commit message here about how to added the codec ids for your device. https://github.com/xbmc/xbmc/pull/22560/commits/f6754bec84520b0ebc75822dbe50cb77f429698a

neo1973 commented on xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp:
```diff
@@ -348,10 +348,28 @@ void CAESinkPipewire::EnumerateDevicesEx(AEDeviceInfoList& list, bool force)
 
-    if (device.m_channels.Count() == 2 && !device.m_streamTypes.empty())
+    // If DTS-HD-MA or TrueHD are configured 8 channels are needed
+    bool hasHBRFormat = std::find_if(device.m_streamTypes.cbegin(), device.m_streamTypes.cend(),
+                                     [](const auto& streamType)
+                                     {
+                                       if (streamType == CAEStreamInfo::STREAM_TYPE_TRUEHD ||
+                                           streamType == CAEStreamInfo::STREAM_TYPE_DTSHD_MA)
+                                         return true;
+
+                                       return false;
+                                     }) != device.m_streamTypes.cend();
```
The correct STL algorithm to use here would be `std::any_of`. Also the lambda can be simplified:
```suggestion
    bool hasHBRFormat = std::any_of(device.m_streamTypes.cbegin(), device.m_streamTypes.cend(),
                                    [](const auto& streamType)
                                    {
                                      return streamType == CAEStreamInfo::STREAM_TYPE_TRUEHD ||
                                             streamType == CAEStreamInfo::STREAM_TYPE_DTSHD_MA;
                                    });
```

fritsch commented:
Thanks @neo1973 - the thing with "correctness" vs. "simplification" vs. "maintainability, we can discuss next time ;-)

sky42src commented:
I can not get DTS-HD High Resolution Audio to play with pipewire passthrough.  There are only a few in my collection, but none of them works.  
checked and working for me with GBM + pipewrie wireplumber config as suggested by @lrusak on mantic
- :MLP FBA:Dolby TrueHD
- :DTS XLL:DTS-HD Master Audio
- :MLP FBA 16-ch:Dolby TrueHD with Dolby Atmos
- :DTS XLL X:DTS-HD MA + DTS:X

not working  
- :DTS XXCH:DTS-HD High Resolution Audio 7.1
- :DTS XBR:DTS-HD High Resolution Audio 5.1

After setup up a baseline with Ubuntu 23.10 and (learning) compiling kodi with GBM my self i did some tests with GBM / Wayland + Pipewire and this PR. Tested on MiniX Z100 with Intel N100.

fritsch commented:
I know that you highjack this PR to fix your local configuration issues, but that's not how it works. This change only fixes a shortcoming in the original implementation to convince kodi to show the format available. It does not change any inner workings.

Also whenever you test or need some help for configuration, without any logs, no support is possible. Also using a setup where "plain ALSA" is properly working with the named codecs is a must. Pipewire cannot do magic.

sky42src commented:
Sorry!  
I thought testing this PR was a good idea.  
Yes Alsa works, but the same with pipewire does not.  
I will leave this PR alone.

fritsch commented:
No, but I give you a choice.

This is github, means: If something is not working, we need every possible logfile, which in this case is kodi.log - but that would most likely be trivial, cause we only see that it is opening 8 channels 192 khz 16 bit with DTS-HD-MA, then we will most likely see some errors in kodi, like AddPacket timed out or something.

Means: pipewire's backend log is needed, so that we can see what is going on.

Summary: Feel free to post, but act like a developer by providing the necessariy information, else everyone can only say: I feel really sorry, that it does not work for you ;-) but this does not help, right?

sky42src commented:
I hear you, but the hacky Ubuntu 23.10 OS installation is no base for a reproducible result with logs to post, but i wanted to say DTS-HD HRA does not work and the rest does.  
I build me a LibreELEC 12 master with runtime switchable audio backends, but the results are strange and need more testing. When i have a clean test case with reproducible results i will post logs for it.  

I only tested in the hope to find a workaround for a Intel GPU gen12 and later audio problem that i stumbled upon last year
https://gitlab.freedesktop.org/drm/intel/-/issues/10199  
but i do know now pipewire does not help.
